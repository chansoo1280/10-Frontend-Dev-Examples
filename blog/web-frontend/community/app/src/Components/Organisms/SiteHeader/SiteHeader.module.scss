@import "@Styles/scss/default.scss";

.site-header {
    position: sticky;
    top: 0;
    background-color: $black;
    color: $white;

    &__title {
        color: $white;
    }

    &__btn {
        background-color: $black;
        color: $white
    }
}