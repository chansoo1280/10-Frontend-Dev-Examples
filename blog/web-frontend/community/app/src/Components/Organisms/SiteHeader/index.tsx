// #region Global Imports
import React, { useEffect, useState } from "react"
import classNames from "classnames"
import Image, { ImageLoader } from "next/image"
// #endregion Global Imports

// #region Local Imports
import { defaultProps, Icon, Space, Button, Text } from "@Components"
import styles from "./SiteHeader.module.scss"
import { useUser } from "@Hooks/useUser"
import { APILogoutGET, Http, ReqType } from "@Services"
import { ResMessageWithDesc, ResStatus } from "@Server/response"
// #endregion Local Imports
const myLoader: ImageLoader = ({ src, width, quality }) => {
    return `/public/Images/${src}?w=${width}&q=${quality || 75}`
}
const SiteHeader = (props: defaultProps): JSX.Element => {
    const { show, className, ...rest } = props
    const { user, clearUser } = useUser()
    const prefixCls = "site-header"
    const classes = classNames(
        styles[`${prefixCls}`],
        {
            [styles[`${prefixCls}--hide`]]: show === false,
        },
        className,
    )
    const logout = async () =>
        await Http<APILogoutGET>(ReqType.GET, ["/api/account/logout"], {}).catch((e: ResMessageWithDesc) => {
            console.log(e)
            switch (e.status) {
                case ResStatus.BadRequest:
                    console.log(e.description)
                    return null

                default:
                    break
            }
            return null
        })

    const handleClickLogout = async () => {
        const user = await logout()
        if (user === null) {
            return
        }
        clearUser()
    }

    return (
        <Space className={classes} padding={user === null ? "4px 4px 4px 16px" : "4px 16px"} {...rest}>
            <Space.Box>
                <Text className={classNames(styles[`${prefixCls}__title`])}>Exeample Project</Text>
            </Space.Box>
            <Button className={classNames(styles[`${prefixCls}__btn`])} size="small" type="text" icon={<Icon iconName="xi-search" />} />
            <Button className={classNames(styles[`${prefixCls}__btn`])} size="small" type="text" icon={<Icon iconName="xi-help-o" />} />

            {user !== null && (
                <>
                    <Button show={user !== null} className={classNames(styles[`${prefixCls}__btn`])} size="small" type="text" icon={<Icon iconName="xi-bell-o" />} />
                    <Button onClick={handleClickLogout} show={user !== null} className={classNames(styles[`${prefixCls}__btn`])} size="small" type="text">
                        <Image loader={myLoader} src="me.png" alt="Picture of the user" width={24} height={24} />
                        {user.name}
                    </Button>
                </>
            )}
            <Button className={classNames(styles[`${prefixCls}__btn`])} size="small" type="text" icon={<Icon iconName="xi-translate" />} />
            <Space show={user === null}>
                <Button href={"/account/login"} size="small">
                    Log in
                </Button>
                <Button href={"/account/register"} className={classNames(styles[`${prefixCls}__btn`])} size="small" type="secondary">
                    Register
                </Button>
            </Space>
        </Space>
    )
}
export default SiteHeader
