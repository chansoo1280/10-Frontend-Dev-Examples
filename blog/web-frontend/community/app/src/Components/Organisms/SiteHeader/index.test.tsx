import { fireEvent, render, screen } from "@testing-library/react"
import SiteHeader from "."
describe("SiteHeader", () => {
    it("renders siteHeader unchanged", () => {
        const { container } = render(<SiteHeader />)
        expect(container).toMatchSnapshot()
    })
    it("should render a siteHeader with the class", () => {
        render(<SiteHeader data-testid="siteHeader" />)

        const siteHeader = screen.getByTestId("siteHeader")
        expect(siteHeader).toBeInTheDocument()
        expect(siteHeader).toHaveClass("space", "site-header")
    })
})
