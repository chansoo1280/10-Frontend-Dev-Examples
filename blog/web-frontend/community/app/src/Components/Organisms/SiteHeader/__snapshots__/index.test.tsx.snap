// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteHeader renders siteHeader unchanged 1`] = `
<div>
  <div
    class="space site-header"
    style="padding: 4px 16px;"
  >
    <div
      class="space__box"
    >
      <span>
        Exeample Project
      </span>
    </div>
    <button
      class="btn btn--small btn--text btn--icon-only site-header__btn"
      data-click-animating="false"
      type="button"
    >
      <i
        class="className xi-icon xi-search"
      />
    </button>
    <button
      class="btn btn--small btn--text btn--icon-only site-header__btn"
      data-click-animating="false"
      type="button"
    >
      <i
        class="className xi-icon xi-help-o"
      />
    </button>
    <button
      class="btn btn--small btn--text btn--icon-only site-header__btn"
      data-click-animating="false"
      type="button"
    >
      <i
        class="className xi-icon xi-bell-o"
      />
    </button>
    <button
      class="btn btn--small btn--text site-header__btn"
      data-click-animating="false"
      type="button"
    >
      <img
        alt="Picture of the user"
        data-nimg="1"
        decoding="async"
        height="24"
        loading="lazy"
        src="/public/Images/me.png?w=48&q=75"
        srcset="/public/Images/me.png?w=32&q=75 1x, /public/Images/me.png?w=48&q=75 2x"
        style="color: transparent;"
        width="24"
      />
      username
    </button>
    <button
      class="btn btn--small btn--text btn--icon-only site-header__btn"
      data-click-animating="false"
      type="button"
    >
      <i
        class="className xi-icon xi-translate"
      />
    </button>
    <div
      class="space space--hide"
    >
      <a
        class="btn btn--small btn--primary"
        href="/login"
      >
        Log in
      </a>
      <button
        class="btn btn--small btn--secondary site-header__btn"
        data-click-animating="false"
        type="button"
      >
        Register
      </button>
    </div>
  </div>
</div>
`;
